be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/Validator.scala

```diff
@@ -60,7 +59,8 @@ object Validator {
     val factory: JsonSchemaFactory = JsonSchemaFactory.getInstance(SpecVersion.VersionFlag.V7)
     // Creation of a JsonNode using the readTree function from the file query.json (at queryPath)
     // FIXME: error handling for queryPath
-    val jsonSchemaNode: JsonNode = objectMapper.readTree(new File(queryPath))
+    val jsonSchemaNode: JsonNode = objectMapper.readTree(queryFile)
+    queryFile.close()
```

Comment:
won't this lead to a failure the next time `validateSchema` is called? 

Also, for readability, it might make sense to create queryFile right before it's used, use it, and close it afterwards.

Reply:
No it does not fail....however it's unnecessary as the `objectMapper.readTree` does the closing.
Additionally, the current state of the method is computationally heavy, meaning to validate every message the validator has to read and process json skeleton from `query.json` (constant) to validate the `jsonString`.  
Calls to readTree are expensive. A fix to this was made in commit `9a6be31fad59817fa369969d26efc60c3217e759` to do this in a nice way. This is a major issue as I have repeatedly noticed a delay of the responses. now the delay is only noticeable in the first message all the others are instant.
A improvement could be computing the schema at compile time for good performance.

Reply:
Thank you for the nice analysis, it seems your new solution is much better indeed (though I haven't looked into all the tiny details) ! 

As opposed to compile-time evaluation, it could also be done on the startup of the server (before the evaluation of the first message), but don't bother with it. There's no need to push the optimization that far ;-)